front: introduce some train endpoints and drop useSpeedSpaceChart

clarani opened the pull request:
For the timetableItem, path and simulation, instead of instantiating 2 queries on the paced train and trainSchedule endpoints, I now instantiate a single query which, depending on the type supplied, then calls the correct endpoint.
Thanks to these new queries, RTK behaviors are preserved, enabling automatic data refetching as soon as a timetableItem is modified.

Also closes #11256

See commits

clarani committed:
front: introduce useSelectedTrain hook and getTimetableItemById endpoint

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: introduce getTrainPath endpoint and use it in useSimulationResults

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: introduce getTrainSimulation endpoint and use it in useSimulationResults

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: drop useSpeedSpaceChart hook

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

emersion commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -1,164 +1,128 @@
+  const [getRollingStockByName] =
+    osrdEditoastApi.endpoints.getRollingStockNameByRollingStockName.useLazyQuery();
+
+  useEffect(() => {
```
Can we avoid `useEffect()` here? See: https://react.dev/learn/you-might-not-need-an-effect

clarani replied:
Done

emersion commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -70,6 +80,105 @@ const osrdEditoastApi = generatedEditoastApi
         },
         providesTags: ['timetable'],
       }),
+      getTimetableItemById: builder.query<TimetableItem | undefined, { id?: TimetableItemId }>({
```
Seems a bit weird to allow undefined IDs as input parameters. Usually this is handled via `skip` or a skip token.

clarani replied:
Done

SharglutDev replied:
For this kind of [case](https://github.com/OpenRailAssociation/osrd/pull/12177#discussion_r2168949662), it would be nice to be able to pass an `undefined` id so we can return `undefined` no ?

emersion replied:
I don't think so, I've replied in that other thread :P

SharglutDev commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -70,6 +77,34 @@ const osrdEditoastApi = generatedEditoastApi
+          }
+          return { data };
+        },
+        providesTags: ['timetable', 'train_schedule', 'paced_train'],
```
Do we need to give tags to this endpoint since we don't subscribe to it ?

clarani replied:
Yes we need to tag this endpoint, since we want to use the default behaviour of RTK.
The data will be re-fetched automatically if a train is updated. I used the tags of `getTrainScheduleById` and `getPacedTrainById`. Maybe we should cleaned the tags later and also use a smarter cache

SharglutDev replied:
Ok, I understand. So that also means that if select train 1 then 2 then 1 again, we won't have any cache and we will refetch the whole train again ?

clarani replied:
Hum no, I think the data will cached in this case since nothing would have invalidated the cache.

But if I select train 1, select train 2, edit train 2, when I will select the train 1 again, it will be fetched again

emersion replied:
I think we should set the provided tags depending on the kind of ID passed in.

SharglutDev replied:
Not sure I understand that from the [docs](https://redux-toolkit.js.org/rtk-query/usage/cache-behavior#default-cache-behavior), but I might be mistaken :)
If we don't subscribe, we don't store anything in the cache so when fetching again it will serve always fresh data.

SharglutDev replied:
Ok got it now (thanks @emersion for the explanation)

SharglutDev replied:
We thought it would be nice to be more granular for the `providedTags` (use the timetable ids), but this can be done in a different PR as it requires to add more endpoints to `osrdEditoastApi` (all the ones invalidating `train_schedule` or `paced_train` tags)

SharglutDev commented on front/src/modules/trainschedule/hooks/useSelectedTrain.tsx:
```diff
@@ -0,0 +1,60 @@
+      .add(occurrenceIndex * pacedTrainIntervalInMs, 'ms')
+      .toISOString();
+    return { ...trainProps, id: trainId, start_time: occurrenceStartTime };
+  }, [trainId, timetableItem]);
```
Since `timetableItem` depends on `timetableItemId` which depends on `trainId`, do we need to add `trainId` to the depths ? This could remove one unnecessary compute for the `useMemo`.

clarani replied:
Hum actually, what I fear is this scenario:
- I have a selected train (the last one of my timetable), so I have a trainId and a timetableItem
- I delete the timetableItem
- then trainId is undefined. In the useQuery, the query is skipped and so we still have the last timetableItem. UseSelectedTrain will then return the previous selected train.

To avoid this case, I think it's better to keep this dependency

SharglutDev replied:
It will but since `useSimulationResults` returns `undefined` if `trainId` is `undefined`, we should be fine no ?

clarani replied:
Yes but what if we use useSelectedTrain at an other place ? I think `useSelectedTrain` should always return the real selected train (and undefined if trainId is undefined).
I can also remove the test in useSimulationResults

emersion replied:
> In the useQuery, the query is skipped and so we still have the last timetableItem. UseSelectedTrain will then return the previous selected train.

This is the default rtk-query behavior, to avoid flashing with empty contents when loading another item. If that's undesirable, the `currentData` field can be used instead of `data`. See https://redux-toolkit.js.org/rtk-query/usage/queries#frequently-used-query-hook-return-values

SharglutDev commented on front/src/modules/trainschedule/hooks/useSelectedTrain.tsx:
```diff
@@ -0,0 +1,60 @@
+
+  return useMemo(() => {
+    if (!trainId || !timetableItem) return undefined;
+    if (isTrainScheduleId(trainId)) {
+      if (isPacedTrainResponseWithPacedTrainId(timetableItem) || trainId !== timetableItem.id)
+        return undefined;
+      return timetableItem;
+    }
```
If we made the previous change in the depths, do we still need this part ?

SharglutDev commented on front/src/modules/trainschedule/hooks/useSelectedTrain.tsx:
```diff
@@ -0,0 +1,60 @@
+    )
+      return undefined;
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { id, exceptions, paced, ...trainProps } = timetableItem;
```
nit : we could also do that to avoid the "eslint disable"
```suggestion
    const { id: _id, exceptions: _exceptions, paced: _paced, ...trainProps } = timetableItem;
```

clarani replied:
Hum I prefer to keep the eslint disable instead of renaming all the unused props with an underscore :)

SharglutDev replied:
Fine by me :)

SharglutDev commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -101,45 +85,18 @@ const useSimulationResults = (infraId: number): SimulationResults | undefined =>
-    const selectedOccurrenceStartTime: string = dayjs(selectedPacedTrain.start_time)
-      .add(selectedOccurrenceIndex * pacedTrainIntervalInMs, 'ms')
-      .toISOString();
+    if (!selectedTrainId || !train) return undefined;
```
If `selectedTrainId` is `undefined`, `train` will be `undefined`. We could probably keep only `train` in the condition

SharglutDev commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -105,6 +107,35 @@ const osrdEditoastApi = generatedEditoastApi
+          }
+          return { data: path };
+        },
+        providesTags: ['pathfinding', 'train_schedule', 'paced_train'],
```
Same as previous comment for relevant tags with `subscribe: false`

SharglutDev commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -136,6 +137,40 @@ const osrdEditoastApi = generatedEditoastApi
+          }
+          return { data: simulation };
+        },
+        providesTags: ['train_schedule', 'paced_train'],
```
ditto

SharglutDev commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -40,37 +42,62 @@ const useSimulationResults = (infraId: number): SimulationResults | undefined =>
+      {
+        skip: !train,
+      }
+    );
```
In this [PR](https://github.com/OpenRailAssociation/osrd/pull/12165), `useScenarioData` will now return the rolling stock list as well. Could we leave a todo here to replace this api call by matching the train rolling stock in this list ?

SharglutDev commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -40,37 +42,62 @@ const useSimulationResults = (infraId: number): SimulationResults | undefined =>
+        },
+      },
+      { skip: path?.status !== 'success' }
+    );
```
We could use `usePathProperties` for this ?

clarani replied:
Ah I would like to remove this hook in the near future, is it ok for you if we keep it like this ?

SharglutDev replied:
Sure ! Forgot you mastermind plan to clean the codebase 😎

SharglutDev commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -40,37 +42,62 @@ const useSimulationResults = (infraId: number): SimulationResults | undefined =>
+    simulation.electrical_profiles,
+    rawPathProperties,
+    path,
+    train.path,
```
This is confusing. Maybe `path` (in the related commit) should be renamed `pathfinding` or `pathfindingResult` ?

clarani replied:
Done :+1:

SharglutDev replied:
Can be done in a different PR but maybe the hook should also return `pathfinding` and not `path` :)

clarani replied:
Noted, I will do this in a future PR :+1:

SharglutDev reviewed (commented):
Thank you for this great refacto !

SharglutDev commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
While testing the PR, got this : 
![Capture d’écran 2025-06-27 à 12 09 14](https://github.com/user-attachments/assets/897a44f5-3adc-4736-bebf-81759b11bef9)

It happened when I had 2 trains left in the timetable : one train schedule and one paced train. I deleted the train schedule and got the white screen.

clarani replied:
Fixed :)

clarani committed:
front: use currentData for simulationResults endpoints

Sometimes, console errors were printed when formatting the pathProperties due
to desynchronization between several endpoints (especially simulation and path
properties).
This bug (#11256) became more critical with my previous commit, as it now breaks
the application and results in blank pages.

To fix this issue, we now use the currentData from the endpoints. This ensures
that while the data is still being fetched, the component handles an undefined
value instead of the previously cached one.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev reviewed (approved):
Lgtm and tested, great job !

emersion commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -70,6 +80,97 @@ const osrdEditoastApi = generatedEditoastApi
+          }
+          return { data: path };
+        },
+        providesTags: ['pathfinding', 'train_schedule', 'paced_train'],
```
I think we can return the right kind of tag depending on the args here:

```ts
providesTags: (result, error, arg, meta) =>
  ['pathfinding', isTrainScheduleId(arg.id) ? 'train_schedule' : 'paced_train'],
```

clarani replied:
Done :)

emersion reviewed (approved):
I haven't looked in depth, but the meat of this PR looks good to me!
